### PR TITLE
PACT-810 message recovery from disk

### PR DESCRIPTION
**Pull request by felixgb**

### CODE REVIEW

File: src/it/scala/MessageRecoveryISpec.scala [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    Config(
+      endpoint = simpleQueueService.Endpoint(Some(DirectAddress(HTTP, sqsHostName, Some(sqsPort))), "elasticmq"),
+      credentials = Some(Credentials("x", "x")),
+      clientId = simpleQueueService.ClientId("ctd-omega-services"),
+      None
+    )
+  )
+
+  private def readTextMessage(jmsMessage: JmsMessage): IO[Unit] = {
+    replyMessageId = jmsMessage.getJMSCorrelationId
```
Comment (rwalpole) [inline src/it/scala/MessageRecoveryISpec.scala:58]: See previous comment regarding `replyMessageId`.

File: src/it/scala/MessageRecoveryISpec.scala [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    apiService.get.start.unsafeToFuture()
+    ()
+  }
+  override def afterAll(): Unit =
+    apiService.get.stop().unsafeToFuture()
+
+  "The Message Recovery API" - {
+
+    "runs the recovery service and removes the message from the message store" in {
+      val messageStoreFolder = Paths.get(apiService.get.config.tempMessageDir)
```
Comment (rwalpole) [inline src/it/scala/MessageRecoveryISpec.scala:104]: You could use `tempMsgDir.get` here intsead of `apiService.get.config.tempMessageDir` as it is the same thing.

File: src/it/scala/MessageRecoveryISpec.scala [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      val localMessageStore = new LocalMessageStore(messageStoreFolder)
+      eventually {
+        localMessageStore.readMessage(messageId).asserting(_.failure.exception mustBe a[NoSuchFileException]) *>
+          IO.pure(replyMessageText).asserting(_ mustBe Some("The Echo Service says: Test World!"))
```
Comment (rwalpole) [inline src/it/scala/MessageRecoveryISpec.scala:108]: You could put the message text in a `val` to show that the reply is the same thing. e.g. `s"The Echo Service says: $testMessage"`

File: src/it/scala/MessageRecoveryISpec.scala [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    jmsMessage.asTextF[IO].attempt.map {
+      case Right(text) =>
+        replyMessageText = Some(text)
+      case Left(e) => fail(s"Unable to read message contents due to ${e.getMessage}")
+    }
+  }
+
+  override protected def beforeAll(): Unit = {
+    // load messages to disk
+    val path = writeMessageFile(new LocalMessage(messageId, "Test World!", None, None, None, None, None, None, None))
```
Comment (rwalpole) [inline src/it/scala/MessageRecoveryISpec.scala:68]: This test is not working currently as the message does not pass validation and specifically is missing a correlationId which means that no reply is sent. I suggest that you copy the `generateValidLocalMessageService` function from the `DispatcherSpec` to here and use that to create your `LocalMessage`. You can change the `messageText` value as required using the copy function.

File: src/test/scala/uk/gov/nationalarchives/omega/api/services/DispatcherSpec.scala [hunk 5]
```diff
@@ -164,11 +166,28 @@ class DispatcherSpec
...
     }
+
+    "for message recovery then it should run recovery and delete all the local messages" in {
+      val mockJmsMessage1 = "Hello, world"
+      val messageId1 = Version1UUID.generate()
+      val mockJmsMessage2 = "Hello, world, again"
+      val messageId2 = Version1UUID.generate()
+
+      writeMessageFile(LocalMessage(messageId1, mockJmsMessage1, None, None, None, None, None, None, None))
+      writeMessageFile(LocalMessage(messageId2, mockJmsMessage2, None, None, None, None, None, None, None))
```
[the remark this answers is not in the document]
Comment (userjd) [reply src/test/scala/uk/gov/nationalarchives/omega/api/services/DispatcherSpec.scala]: Done

File: src/it/scala/MessageRecoveryISpec.scala (hunk 3 above)
Comment (userjd) [reply src/it/scala/MessageRecoveryISpec.scala]: Done

File: src/it/scala/MessageRecoveryISpec.scala (hunk 4 above)
Comment (userjd) [reply src/it/scala/MessageRecoveryISpec.scala]: Done

File: src/it/scala/MessageRecoveryISpec.scala (hunk 2 above)
Comment (userjd) [reply src/it/scala/MessageRecoveryISpec.scala]: Done

File: src/it/scala/MessageRecoveryISpec.scala (hunk 1 above)
Comment (userjd) [reply src/it/scala/MessageRecoveryISpec.scala]: Done

File: src/it/scala/MessageRecoveryISpec.scala [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+  private val requestQueueName = "request-general"
+  private val replyQueueName = "omega-editorial-web-application-instance-1"
+  private val sqsHostName = "localhost"
+  private val sqsPort = 9324
+
+  private val messageId = Version1UUID.generate()
+
+  var replyMessageText: Option[String] = None
+  var replyMessageId: Option[String] = None
```
[the remark this answers is not in the document]
Comment (userjd) [reply src/it/scala/MessageRecoveryISpec.scala]: Done

**Review by rwalpole — APPROVED** [review]

**Pull request merged**